OpenRA.Mods.AS/Activities/EnterNydus.cs

```diff
@@ -0,0 +1,110 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2015- OpenRA.Mods.AS Developers (see AUTHORS)
+ * This file is a part of a third-party plugin for OpenRA, which is
+ * free software. It is made available to you under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation. For more information, see COPYING.
+ */
+#endregion
+
+using System.Drawing;
+using OpenRA.Mods.AS.Traits;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.AS.Activities
+{
+	class EnterNydus : Enter
+	{
+		public EnterNydus(Actor self, Actor target, EnterBehaviour enterBehaviour)
+			: base(self, target, enterBehaviour)
+		{
+		}
+
+		protected override bool CanReserve(Actor self)
+		{
+			// Primary building is where you come out!
+			return !Target.Actor.IsPrimaryNydusExit();
+		}
+
+		protected override void OnInside(Actor self)
+		{
+			// entered the nydus canal but the entrance is dead immediately. haha;;
+			if (Target.Actor.IsDead)
+				return;
+
+			// Find the primary nydus exit.
+			var pri = self.Owner.PlayerActor.Trait<NydusCounter>().PrimaryActor;
```

Comment:
Trait<>() lookups are bad in practice, because so far there isn't a hard rule about the requirement - a Desc mention is NOT such one. Lint rules tend to be one, but I have a better idea.

Also, the logic could very well get a Type selection to support multiple kindofs in which case this would be a First/FirstOrDefault.